### PR TITLE
BUG: Fix wrong reading sparse matrix

### LINKED ISSUE #29814: Pandas wrongly read Scipy sparse matrix
#### Code Sample, a copy-pastable example if possible

```python
>>> from scipy.sparse import coo_matrix
>>> sparse_data = coo_matrix((
...     [0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0],
...     ([0, 1, 1, 2, 2, 3, 3], [0, 1, 2, 0, 2, 0, 1])
... ))
>>> sparse_data.todense()
matrix([[0., 0., 0.],
        [0., 1., 1.],
        [1., 0., 1.],
        [1., 1., 0.]])
>>> pd.DataFrame.sparse.from_spmatrix(sparse_data)
     0    1    2
0  0.0  0.0  0.0
1  0.0  1.0  1.0
2  0.0  0.0  1.0
3  1.0  1.0  0.0
>>> pd.DataFrame.sparse.from_spmatrix(sparse_data.tocsr())
     0    1    2
0  0.0  0.0  0.0
1  0.0  1.0  1.0
2  0.0  0.0  1.0
3  1.0  1.0  0.0
```

#### Problem description

`sparse_data.todense()` and the other matrixes should be same, but not.
 (i.e. the [2, 0] value in bottom 2 matrixes should be 1.)

#### Expected Output

When converting sparse matrix to Pandas DataFrame values of the sparse array should stay the same.

#### Output of ``pd.show_versions()``

<details>

python           : 3.7.5.final.0
pandas           : 0.25.3
numpy            : 1.17.4
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 41.6.0
Cython           : None
pytest           : 5.3.0
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.2
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by m7142yosuke**
- [x] closes #29814 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by dwhswenson** [comment]
Previously the comment in #29814 had suggested using `data.indices` instead of `data.nonzero()`. However, this PR instead strips any explicit zeros in the sparse input.

One of my use cases involves taking the difference of two sparse matrices, and then passing the result to scipy.sparse and pandas (ideally filling missing with NaN, instead of 0). There's a meaningful difference between "this element has no information from either matrix" (NaN) and "this element had the exact same value in both matrices" (0). It looks to me like this PR would obscure that distinction.

This would also make it hard to round-trip from `scipy.sparse`, which seems to handle explicit zeros.

Is there a solution that doesn't involve losing the information about explicit zeros?

(This wouldn't be a show-stopper for us; right now export to pandas is basically a convenience for users who want it. We'd just have to add a warning on that function.)

**Comment by WillAyd** [comment]
@jorisvandenbossche or @TomAugspurger any thoughts here?

**Comment by TomAugspurger** [comment]
What’s the performance cost of your PR be using indicies? It seems like stripping explicit zeros will require scanning the full data?

**Comment by m7142yosuke** [comment]
@TomAugspurger 
Eliminating explicit zeros will require scanning the full data in my PR.
I didn't come up with other way(not scanning the full data). Is there another good way?

**Comment by jorisvandenbossche** [comment]
@m7142yosuke as I also mentioned on the issue, there has been a PR that refactored the `from_spmatrix` dataframe accessor method: https://github.com/pandas-dev/pandas/pull/32825

This seems to have fixed the DataFrame bug (it has its own logic now, and does not use SparseArray.from_spmatrix under the hood), but so the SparseArray is still buggy (what you are fixing here). 
But so, it's probably worth to take a look at how that PR changed the implementation, and do something similar here (or try to see if some of the code can be shared).

Also adding a test case for the original case from https://github.com/pandas-dev/pandas/issues/29814 would be welcome

**Review by jreback — CHANGES_REQUESTED** [review]
can you update to the comments on the PR, e.g. @jorisvandenbossche mentioned an explicit test & add a whatsnew note in 1.1., sparse bug fix section. pls merge master and ping on green.

**Comment by m7142yosuke** [comment]
@jorisvandenbossche @jreback 
Please review this PR

Done.
- Refactoring code referencing #32825
- Add a test case for the original case from #29814

File: doc/source/whatsnew/v1.1.0.rst [hunk 1]
```diff
@@ -572,7 +572,7 @@ Reshaping
 Sparse
 ^^^^^^
 - Creating a :class:`SparseArray` from timezone-aware dtype will issue a warning before dropping timezone information, instead of doing so silently (:issue:`32501`)
--
+- Bug in :meth:`SparseArray.from_spmatrix` wrongly read scipy sparse matrix (:issue:`31991`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.1.0.rst:575]:
```suggestion
- Bug in :meth:`arrays.SparseArray.from_spmatrix` wrongly read scipy sparse matrix (:issue:`31991`)
```

**Review by TomAugspurger — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @m7142yosuke

**Pull request merged**